lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -4193,23 +4317,98 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
+      //    The adjoint successors correspond to the predecessors of the current
+      //    block.
       SmallVector<std::pair<EnumElementDecl *, SILBasicBlock *>, 4>
-          predecessorCases;
+          adjointSuccessorCases;
       for (auto *predBB : bb->getPredecessorBlocks()) {
-        auto *adjointSuccBB = getAdjointBlock(predBB);
+        // Get the adjoint block and optional adjoint trampoline block of the
+        // predecessor block.
+        auto *adjointBB = getAdjointBlock(predBB);
+        auto *adjointTrampolineBB = getAdjointTrampolineBlock(predBB, bb);
+        SILBasicBlock *adjointSuccBB = nullptr;
+        // If the predecessor block does not have a corresponding adjoint
+        // trampoline block, then the adjoint successor is the adjoint block.
+        if (!adjointTrampolineBB) {
+          adjointSuccBB = adjointBB;
+        }
+        // Otherwise, the adjoint successor is the adjoint trampoline block,
+        // which branches to the adjoint block and propagates adjoint values of
+        // active values.
+        else {
+          adjointSuccBB = adjointTrampolineBB;
+          assert(adjointSuccBB && adjointSuccBB->getNumArguments() == 1);
+          SILBuilder adjointTrampolineBBBuilder(adjointSuccBB);
+          SmallVector<SILValue, 8> trampolineArguments;
+          // Propagate pullback struct argument.
+          trampolineArguments.push_back(adjointSuccBB->getArguments().front());
+          // Propagate adjoint values/buffers of active values/buffers.
+          for (unsigned i : indices(activeValues[predBB])) {
+            auto activeValue = activeValues[predBB][i];
+            if (activeValue->getType().isObject()) {
+              auto activeValueAdj = getAdjointValue(bb, activeValue);
+              auto concreteActiveValueAdj =
+                  materializeAdjointDirect(activeValueAdj, adjLoc);
+              trampolineArguments.push_back(concreteActiveValueAdj);
+              // If the adjoint block does not yet have a registered adjoint
+              // value for the active value, set the adjoint value to the
+              // forwarded adjoint value argument.
+              // TODO: Hoist this logic out of loop over predecessor blocks to
+              // remove the `hasAdjointValue` check.
+              if (!hasAdjointValue(predBB, activeValue)) {
+                auto *adjointBBArg =
+                    getActiveValueAdjointBlockArgument(predBB, activeValue);
+                auto forwardedArgAdj =
+                    makeConcreteAdjointValue(ValueWithCleanup(adjointBBArg));
+                initializeAdjointValue(predBB, activeValue, forwardedArgAdj);
+              }
+            } else {
+              // Propagate adjoint buffers using `copy_addr`.
+              auto adjBuf = getAdjointBuffer(bb, activeValue);
+              auto succAdjBuf = getAdjointBuffer(predBB, activeValue);
+              builder.createCopyAddr(
```

Comment:
We should put more consideration into cleanup handling. The current implementation may cause segfaults. When you copy something to a function-local buffer, you should _transfer_ the ownership of the cleanup to the destination buffer so that the underlying value only gets cleaned up when the buffer is getting deallocated.

Reply:
I see. Is this the correct logic or no?
```cpp
              auto adjBuf = getAdjointBuffer(bb, activeValue);
              auto succAdjBuf = getAdjointBuffer(predBB, activeValue);
              succAdjBuf.setCleanup(makeCleanupFromChildren(
                  {adjBuf.getCleanup(), succAdjBuf.getCleanup()}));
```

Reply:
To step back, there are two problems:
1. How do you guarantee `activeValue` has never been an argument to `getAdjointBuffer(predBB, activeValue)` before? That is, will there be a case where `getAdjointBuffer(predBB, activeValue)` is called on the same `activeValue` twice? If that does happen, this code will simply rewrite the the existing adjoint value, which may give you a miscalculation error. Otherwise, it's worth thinking about how the code can be structured in a way that the reader understands this is the first time such a buffer is created for this particular adjoint value.
2. Back to cleanups, the way you should deal with cleanups is directly related to (1). If there are multiple calls to `getAdjointBuffer(predBB, activeValue)` for the same `activeValue`, then you should probably do an `AdditiveArithmetic.+=(&succAdjBuf, adjBuf)` and transfer no cleanup. Otherwise, do `succAdjBuf.setCleanup(adjBuf.getCleanup())` and set `adjBuf`'s cleanup to `nullptr`.

In any case, I think (1) should be resolved first.

Reply:
We peer programmed and discussed the two problems in depth.

We found that `getAdjointBuffer(predBB, activeValue)` is called multiple times for the same `activeValue`. But performing `AdditiveArithmetic.+=(&predAdjBuf, adjBuf)` instead of `copy_addr(adjBuf, predAdjBuf)` lead to computation errors.

Progress in https://github.com/apple/swift/pull/25057/commits/49937d93f87989f9d46d29a956bb42923ca143d2: `var` + control flow tests are fixed, but cleanups during adjoint value/buffer propagation are not handled, leading to memory leaks.

Negative leak checking tests added to `test/AutoDiff/leakchecking.swift`. @rxwei offered to help fix the cleanup logic in a follow-up PR.